### PR TITLE
[Fixes #4043] Figure out ControllerModel.AttributeRoutes

### LINKED ISSUE #4043: Figure out ControllerModel.AttributeRoutes
`ControllerModel` supports multiple routes per controller, but `ActionModel` does not.

If you have a controller with multiple `[Route]` attributes, this results in a single `ControllerModel` instance.

If you have an action with multiple `[Route]` attributes, this results in multiple `ActionModel` instances.

We should resolve this asymmetry somehow.

### PR DESCRIPTION
**Pull request by kichalla**
@rynowak

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ControllerModel.cs [hunk 1]
```diff
@@ -6,14 +6,13 @@
...
 using System.Reflection;
-using Microsoft.AspNetCore.Mvc.ActionConstraints;
 using Microsoft.AspNetCore.Mvc.Filters;
 using Microsoft.AspNetCore.Mvc.Routing;
 
 namespace Microsoft.AspNetCore.Mvc.ApplicationModels
 {
     [DebuggerDisplay("Name={ControllerName}, Type={ControllerType.Name}," +
-                     " Routes: {AttributeRoutes.Count}, Filters: {Filters.Count}")]
+                     " Filters: {Filters.Count}")]
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ControllerModel.cs]: Let's simplify this to just the controller type name as well. These DD's aren't super readable as-is.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs [hunk 2]
```diff
@@ -185,67 +185,87 @@ public static IList<ControllerActionDescriptor> Build(ApplicationModel applicati
...
+                    actionDescriptors.Add(actionDescriptor);
+
+                    IList<IActionConstraintMetadata> controllerConstraints = null;
+                    if (controller.Selectors.Count > 0)
+                    {
+                        controllerConstraints = controller.Selectors[0].ActionConstraints;
+                    }
+
+                    // If there's no attribute route on the controller, then we can use all of the filters/constraints
+                    // on the controller.
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionDescriptorBuilder.cs]: We use all the filters on the controller regardless.

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/DefaultApplicationModelProviderTest.cs [hunk 3]
```diff
@@ -596,30 +618,35 @@ public void BuildActionModels_AttributeRouteOnAction_CreatesOneActionInforPerRou
...
+            foreach (var selectorModel in action.Selectors)
             {
-                Assert.Equal("Index", action.ActionName);
-                Assert.NotNull(action.AttributeRouteModel);
+                Assert.NotNull(selectorModel.AttributeRouteModel);
             }
 
-            var list = Assert.Single(actions, ai => ai.AttributeRouteModel.Template.Equals("List"));
-            var listMethod = Assert.Single(list.HttpMethods);
+            var selModel = Assert.Single(action.Selectors, ai => ai.AttributeRouteModel?.Template == "List");
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/DefaultApplicationModelProviderTest.cs]: Don't abbreviate

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/DefaultApplicationModelProviderTest.cs [hunk 4]
```diff
@@ -627,77 +654,76 @@ public void BuildActionModels_NoRouteOnController_AllowsConventionallyRoutedActi
...
             var builder = new TestApplicationModelProvider();
             var typeInfo = controller.GetTypeInfo();
 
             // Act
-            var actions = builder.BuildActionModels(typeInfo, typeInfo.AsType().GetMethod("Index"));
+            var action = builder.CreateActionModel(typeInfo, typeInfo.AsType().GetMethod("Index"));
 
             // Assert
-            Assert.Equal(2, actions.Count());
+            //TODO: verify the attribute types
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/DefaultApplicationModelProviderTest.cs]: No TODOs

**Comment by rynowak** [comment]
:shipit: after fixing minor comments - looks good 

Can you open a follow up to add docs to application model?

**Comment by kichalla** [comment]
Thanks. Filed a doc issue now: https://github.com/aspnet/Mvc/issues/4199

**Pull request closed without merging**